### PR TITLE
[sui-framework] separate storage fund into two parts

### PR DESCRIPTION
**Pull request by emmazzz**
## Description 

This PR splits up the storage fund into two `Balance`s and puts it into its own module. 
The two balances are `total_object_storage_rebates` and `non_refundable_balance` (I'm sure better names exist and happy to take suggestions).  
`total_object_storage_rebates` maintains the invariant that`\forall o, sum(o.storage_rebate) = storage_fund.total_object_storage_rebates.value()` @sblackshear.
And `non_refundable_balance` has everything else that should not be withdrawn unless the system changes its mind.

## Test Plan 

modified existing tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [x] breaking change for FNs (FN binary must upgrade)
- [x] breaking change for validators or node operators (must upgrade binaries)
- [x] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-framework/packages/sui-system/sources/sui_system_state_inner.move [hunk 1]
```diff
@@ -882,7 +896,11 @@ module sui_system::sui_system_state_inner {
     }
 
     public(friend) fun get_storage_fund_balance(self: &SuiSystemStateInner): u64 {
```
Comment (lxfind) [inline crates/sui-framework/packages/sui-system/sources/sui_system_state_inner.move:898]: Probably should be renamed to have "total" in it as well

**Review by lxfind — APPROVED** [review]
LGTM
@dariorussi to double check

**Review by dariorussi — APPROVED** [review]
it looks good. I am going to use `non_refundable_storage_fee` as suggested in the thread. Just fyi

File: crates/sui-adapter/src/execution_engine.rs [hunk 2]
```diff
@@ -455,6 +458,7 @@ fn advance_epoch<S: BackingPackageStore + ParentSync + ChildObjectResolver>(
         storage_charge: change_epoch.storage_charge,
         computation_charge: change_epoch.computation_charge,
         storage_rebate: change_epoch.storage_rebate,
+        non_refundable_storage_rebate: change_epoch.non_refundable_storage_rebate,
```
[the remark this answers is not in the document]
Comment (emmazzz) [reply crates/sui-adapter/src/execution_engine.rs:461]: `change_epoch.non_refundable_storage_rebate` comes from the corresponding field in the epoch rolling gas cost summary (which you added) in the last checkpoint in the epoch.

File: crates/sui-types/src/sui_system_state/sui_system_state_summary.rs [hunk 3]
```diff
@@ -39,6 +42,8 @@ pub struct SuiSystemStateSummary {
     pub safe_mode_computation_rewards: u64,
     /// Amount of storage rebates accumulated (and not yet burned) during safe mode.
     pub safe_mode_storage_rebates: u64,
+    /// Amount of non-refundable storage rebates accumulated during safe mode.
+    pub safe_mode_non_refundable_storage_rebates: u64,
```
[the remark this answers is not in the document]
Comment (emmazzz) [reply crates/sui-types/src/sui_system_state/sui_system_state_summary.rs:46]: Well the whole point of your PR and mine is to keep these two things separate?

File: crates/sui-core/src/authority.rs [hunk 4]
```diff
@@ -3288,6 +3288,8 @@ impl AuthorityState {
             gas_cost_summary.storage_cost,
             gas_cost_summary.computation_cost,
             gas_cost_summary.storage_rebate,
+            // TODO: use the actual non refundable storage rebate from gas_cost_summary
+            0,
```
Comment (emmazzz) [inline crates/sui-core/src/authority.rs:3292]: Btw @dariorussi this where I'll be using the field you are adding.

**Pull request merged**